services/docker/docker-version.service.js

```diff
@@ -0,0 +1,143 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const { nonNegativeInteger } = require('../validators')
+const { latest, renderVersionBadge } = require('../version')
+const {
+  buildDockerUrl,
+  getDockerHubUser,
+  getMultiPageData,
+  getDigestSemVerMatches,
+} = require('./docker-helpers')
+const { NotFound } = require('..')
+const { BaseJsonService } = require('..')
+
+const buildSchema = Joi.object({
+  count: nonNegativeInteger.required(),
+  results: Joi.array().items(
+    Joi.object({
+      name: Joi.string().required(),
+      images: Joi.array().items(
+        Joi.object({
+          digest: Joi.string(),
+          architecture: Joi.string().required(),
+        })
+      ),
+    })
+  ),
+}).required()
+
+const queryParamSchema = Joi.object({
+  sort: Joi.string()
+    .valid('date', 'semver')
+    .default('date'),
+}).required()
+
+module.exports = class DockerVersion extends BaseJsonService {
+  static get category() {
+    return 'version'
+  }
+
+  static get route() {
+    return { ...buildDockerUrl('v', true), queryParamSchema }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Docker Image Version (latest by date)',
+        pattern: ':user/:repo',
+        namedParams: { user: '_', repo: 'alpine' },
+        queryParams: { sort: 'date' },
+        staticPreview: this.render({ version: '3.9.5' }),
+      },
+      {
+        title: 'Docker Image Version (latest semver)',
+        pattern: ':user/:repo',
+        namedParams: { user: '_', repo: 'alpine' },
+        queryParams: { sort: 'semver' },
+        staticPreview: this.render({ version: '3.11.3' }),
+      },
+      {
+        title: 'Docker Image Version (tag latest semver)',
+        pattern: ':user/:repo/:tag',
+        namedParams: { user: '_', repo: 'alpine', tag: '3.6' },
+        staticPreview: this.render({ version: '3.6.5' }),
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return {
+      label: 'version',
+      color: 'blue',
+    }
+  }
+
+  static render({ version }) {
+    return renderVersionBadge({ version })
+  }
+
+  async fetch({ user, repo, page }) {
+    page = page ? `&page=${page}` : ''
+    return this._requestJson({
+      schema: buildSchema,
+      url: `https://registry.hub.docker.com/v2/repositories/${getDockerHubUser(
+        user
+      )}/${repo}/tags?page_size=100&ordering=last_updated${page}`,
+      errorMessages: { 404: 'repository or tag not found' },
+    })
+  }
+
+  transform({ tag, sort, data, pagedData }) {
+    let version
+
+    if (!tag && sort === 'date') {
+      version = data.results[0].name
+      if (version !== 'latest') {
+        return { version }
+      }
+      const { digest } = data.results[0].images.find(
+        i => i.architecture === 'amd64'
+      ) // Digest is the unique field that we utilise to match images
```

Comment:
I've noted the comments in https://github.com/badges/shields/pull/4562#discussion_r370467990 about why `digest` is not required in the schema, but given it is not required, what will happen in the edge case where we could end up matching an object where `digest` is undefined? Do we need a guard condition to throw an `InvalidResponse` in that case?

Reply:
So from what I can tell this could occur on the version badge if someone tries to look up an explicit tag on some library based repos, namely images from < 2015-2016.
It appears that by about June 2016 that the digest is present, perhaps due to a change in spec from Docker.

Guarding for that case makes perfect sense, however, I think instead of returning an invalid response we should return the actual version/tag the user has looked up.

Reply:
Good spot Chris! 

Does the check added to the else block below in https://github.com/badges/shields/pull/4562/commits/0ca1c51cef3b6d111f03d23408b7d84e8508d997 address the scenario (`date` sorting, no specified tag, and the most recent tag by last_updated is `latest`)?

If `digest` is undefined here, are there any circumstances where `getDigestSemVerMatches` would return an undefined/falsy value for `version` (I suppose any value of `data` that would result in `matches` being an empty array)

https://github.com/badges/shields/blob/0ca1c51cef3b6d111f03d23408b7d84e8508d997/services/docker/docker-helpers.js#L45-L56

Reply:
It doesn't and I think that's a very specific edge case which we probably don't need to address.

The case added in https://github.com/badges/shields/commit/0ca1c51cef3b6d111f03d23408b7d84e8508d997 treats an explicit lookup for a tag that doesn't have a digest.

Anything created within the last 4 years will have a digest present, I don't believe you can actually upload an image now without it resulting in a digest being published. I have a feeling this has to do with Registry v1 -> v2 though haven't had much time to look in detail.

Reply:
> It doesn't and I think that's a very specific edge case which we probably don't need to address

While I agree it's an edge case, we should still be cognizant of known possibilities. In my experience with Shields, individual use cases and patterns that seem highly unlikely to occur end up actually happening/being reported a lot more often than you'd think 😄 

For reference, a badge creation attempt with an undefined `message` value would cause a runtime exception. We're using the `renderVersionBadge` helper here that wouldn't happen since the message would end up being a `v` in the worst case.

I'd argue it would be better to show `latest` instead of `v`, and that could be handled very easily here by changing the line below from:
```js
return { version: getDigestSemVerMatches({ data: pagedData, digest }) }
```

to 
```js
return { version: getDigestSemVerMatches({ data: pagedData, digest }) || 'latest' }
```